extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java

```diff
@@ -0,0 +1,88 @@
+
+    /**
+     * To set this, add `-J-Dorg.quarkus.caffeine.graalvm.diagnostics=true` to the native-image parameters
+     */
+    private static final boolean log = Boolean.getBoolean("org.quarkus.caffeine.graalvm.diagnostics");
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        Class<?> caffeineCoreClazz = access.findClassByName("com.github.benmanes.caffeine.cache.Caffeine");
+        access.registerReachabilityHandler(this::ensureCaffeineSupportEnabled, caffeineCoreClazz);
+    }
+
+    private void ensureCaffeineSupportEnabled(DuringAnalysisAccess duringAnalysisAccess) {
+        final boolean needsEnablingYet = triggered.compareAndSet(false, true);
+        if (needsEnablingYet) {
+            if (log) {
+                System.out.println(
+                        "Quarkus's automatic feature for GraalVM native images: enabling support for core Caffeine caches");
+            }
+            registerCaffeineReflections(duringAnalysisAccess);
+        }
+    }
+
+    private void registerCaffeineReflections(DuringAnalysisAccess duringAnalysisAccess) {
+        final String[] needsHavingSimpleConstructors = typesNeedingConstructorsRegistered();
+        for (String className : needsHavingSimpleConstructors) {
+            registerForReflection(className, duringAnalysisAccess);
+        }
+    }
+
+    private void registerForReflection(
+            String className,
+            DuringAnalysisAccess duringAnalysisAccess) {
+        final Class<?> aClass = duringAnalysisAccess.findClassByName(className);
+        final Constructor<?>[] z = aClass.getDeclaredConstructors();
+        RuntimeReflection.register(aClass);
+        RuntimeReflection.register(z);
+    }
+
+    public static String[] typesNeedingConstructorsRegistered() {
+        return new String[] {
+                //N.B. this list is not complete, but a selection of the types we expect being most useful.
+                //unfortunately registering all of them has been shown to have a very significant impact
+                //on executable sizes. See https://github.com/quarkusio/quarkus/issues/12961
+                "com.github.benmanes.caffeine.cache.PDMS",
+                "com.github.benmanes.caffeine.cache.PSA",
+                "com.github.benmanes.caffeine.cache.PSMS",
+                "com.github.benmanes.caffeine.cache.PSW",
+                "com.github.benmanes.caffeine.cache.PSWMS",
+                "com.github.benmanes.caffeine.cache.PSWMW",
+                "com.github.benmanes.caffeine.cache.SILMS",
+                "com.github.benmanes.caffeine.cache.SSA",
+                "com.github.benmanes.caffeine.cache.SSLA",
+                "com.github.benmanes.caffeine.cache.SSLMS",
+                "com.github.benmanes.caffeine.cache.SSMS",
+                "com.github.benmanes.caffeine.cache.SSMSA",
+                "com.github.benmanes.caffeine.cache.SSMSW",
+                "com.github.benmanes.caffeine.cache.SSW",
+        };
```

Comment:
N.B. the list has grown slightly, as I've merged the missing ones from the (now deleted) duplicate registrations that were being triggered by https://github.com/quarkusio/quarkus/pull/17748/files#diff-b8f55d2a209ec481eaa58dee37f596a382aa4443a9585ae00581fcfac9a9c8b3